File: 849/maximize-distance-to-closest-person.js
/**
 * @param {number[]} seats
 * @return {number}
 */
function maxDistToClosest(seats) {
    let len = seats.length;
    let res = []; let dist;
    for (let i = 0; i < len; i++) {
        if (seats[i] == 1)
            continue;
        if (seats[i] != 0 && seats[i] != 1) {
            console.log("the input should only contain 0 and 1!");
        }

        let [left, right] = [0, 0];
        if (i > 0) {
            while (seats[i - left] == 0) {
                left++;
            }
        }

        if (i < len - 1) {
            while (seats[i + right] == 0) {
                right++;
            }
        }
        if (left == 0) {
            dist = right;
        } else if (right == 0) {
            dist = left;
        } else {
            dist = left > right ? right : left;
        }
        res.push(dist);
    }
    return Math.max(...res);
};

File: 849/MaximizeDistanceToClosestPerson.scala
object MaximizeDistanceToClosestPerson {
  def maxDistToClosest(seats: Array[Int]): Int = {
    var res = List.empty[Int]
    val len = seats.length
    for (i <- 0 until len if seats(i) != 1) {
      var (left, right) = (0, 0)
      var dist = 0
      if (i > 0)
        while (seats(i - left) == 0)
          left += 1
      if (i < len - 1)
        while (seats(i + right) == 0)
          right += 1

      if (left == 0) dist = right
      else if (right == 0) dist = left
      else dist = if (left > right) right else left
      res = res :+ dist
    }
    res.max
  }

  def main(args: Array[String]): Unit = {
    val seats = Array(1, 0, 0, 0, 1, 0, 1)
    println(maxDistToClosest(seats))
  }

}
